### ISSUE TITLE
Feature Request: Auto Gradient Calculation

### ISSUE THREAD

**Issue by Ailanz** [deeplearning4j/deeplearning4j#3367]
Hi I'm trying to implement Neural Style in DL4J. I'm able to implement most of it but when trying to optimize a function w.r.t. loss, I'm having trouble implementing the gradient calculations. I am requesting a feature that acts similar to Keras' backend.gradient function: https://keras.io/backend/

Thanks!

**Comment by AlexDBlack** [comment]
This is a pretty huge (as in, a lot of work to implement) feature request. But: it's something we are looking at. It could take quite a while, but we'll be sure to announce it when it is finally ready.
https://github.com/deeplearning4j/nd4j/pull/1750

**Comment by andresmasegosa** [comment]
Same here, Auto Gradient calculation would be a killer feature for me. Which is the priority of this feature request? Any idea about when it will be ready?

**Comment by AlexDBlack** [comment]
@andresmasegosa It's being a actively worked on currently (though it's not the only thing we have going on at the moment).
We don't have an ETA for this yet. At this point, the only thing we can say: it'll be more than a few weeks until it's finished and properly tested/debugged and ready, possibly much more.

**Comment by raver119** [comment]
Implemented as SameDiff and available in 1.0.0-alpha & current master
